File: Java Programs/Java BinaryTree/BinaryTree.java
class BinaryTree <E extends Comparable<E>>
{

 private Node<E> root;

 public BinaryTree()
 {
   root = null;
 }

 public Node<E> getRoot()
 {
   return root;
 }

 public void addNode(E input)
 {

  Node<E> newnode = new Node<E>();
  newnode.value = input;
  newnode.right = null;
  newnode.left = null;

  root = insertNode(root, newnode);
 }

 //returning everything to keep track of the root
 // The tree will construct by using a copy of the root and a copy of newnode. The reason the tree works is because the Tree is constructed, and returned in a manner in which 
 // the trees root is kept track of and the root of the tree will be returned, making the root of the original bst assigned the tree's root. 
 private Node<E> insertNode(Node<E> currentParent, Node<E> newnode)
 {
   if (currentParent == null)
   return newnode;

   else if (newnode.value.compareTo(currentParent.value) < 0)
   currentParent.left = insertNode(currentParent.left, newnode);
 
   else
   currentParent.right = insertNode(currentParent.right, newnode);

  return currentParent;
 }

/*
 public boolean insertNode(E input)
 {
   if (root == null)
   {
     Node<E> newnode = new Node();
     newnode.value = input;
     newnode.right = null;
     newnode.left = null;
     
     root = newnode;
   }

  Node<E> parent = null;
  Node<E> current = root;

 while (current != null)
 {
   if (input.compareTo(current.value) < 0)
   {
     parent = current;
     current = current.left;
   }
   else if (input.compareTo(current.value) > 0)
   {
     parent = current;
     current = current.right;
   }
   else
   return false;
 }

  if (input.compareTo(parent.value) < 0)
  {
    Node<E> newnode = new Node();
    newnode.value = input;
    newnode.right = null;
    newnode.left = null;
     
     parent.left = newnode;
  }
  
  else
  {
    Node<E> newnode = new Node();
    newnode.value = input;
    newnode.right = null;
    newnode.left = null;
     
     parent.right = newnode;
  }

return true;
}
*/
 public void inorder()
 {
   inordertraversal(getRoot());
 }

 private void inordertraversal(Node<E> root)
 {
   if (root == null)
   return;
   inordertraversal(root.left);
   System.out.println(root.value);
   inordertraversal(root.right);
 }

 public void deleteNode(E input)
 {

   Node<E> current = root;
   Node<E> parent = null;
   boolean flag = findNode(input, root);

   if (flag == false)
   {
     System.out.println(input + " is not in the tree");
     return;
   }

   else
   {
	while(current != null)
        {
          if ((current.value).compareTo(input) < 0)
          {
	    parent = current;
	    current = current.right;
          }

          else if ((current.value).compareTo(input) > 0)
          {
            parent = current;
            current = current.left;
          }

          else if (current.value == input)
          break;
         }
    }

 if (current.left == null)
 {
   if (parent == null)
   parent = current.right;

   else
   {
     if (input.compareTo(parent.value) > 0)
     parent.right = current.right;
     else
     parent.left = current.right;
   }
 }

 else
 {
   Node<E> parentOfRightMost = current;
   Node<E> rightMost = current.left;

   while(rightMost.right != null)
   {
	parentOfRightMost = rightMost;
	rightMost = rightMost.right;
   }

   current.value = rightMost.value;

   if (parentOfRightMost.right == rightMost)
   {
	parentOfRightMost.right = rightMost.left;
   }

   else
   {
     parentOfRightMost.left = rightMost.left;
   }
 }
}

 public boolean findNode(E input, Node<E> root)
 {
   while (root != null)
   { 
     if (root.value == input)
     {
    	return true;
     }
     else if (input.compareTo(root.value) > 0)
     {
      root = root.right;
     }
     else
     {
     root = root.left;
     }
  }

  return false;
 }

 /*
 //recursion didnt work for some reason????
 //kept "missing return statement" for no reason?????
 public boolean findNode(E input, Node<E> root)
 {
  if (root != null)
  { 
     if (root.value == input)
     {
    	return true;
     }

     else if (input.compareTo(root.value) > 0)
     {
     findNode(input, root.right);
     }
     else
     {
     findNode(input, root.left);
     }
  }

  else
  return false;
 }
*/
}

File: C++ Programs/C++ BinaryTree/BinaryTree.cpp

#include <stddef.h>
#include "BinaryTree.h"
#include <iostream>
using std::cout;
using std::cin;
using std::endl;

template <typename T>
BinaryTree<T>::BinaryTree()
{

root = NULL;

}

template <typename T>
void BinaryTree<T>::addNode(T input)
{

 Node* newnode = new Node();
 newnode->value = input;
 newnode->left = NULL;
 newnode->right = NULL;

 insertNode(root, newnode);
}

template <typename T>
void BinaryTree<T>::insertNode(Node*& root, Node* & newnode)
{

 if (root == NULL)
    root = newnode;

 else
 {
  if (newnode->value < root->value)
     insertNode(root->left, newnode);
  else
     insertNode(root->right,newnode);
 }
}

template <typename T>
void BinaryTree<T>::inordertraversal(Node* & root)
{
  if (root != NULL)
  {
  	inordertraversal(root->left);
  	cout << root->value << endl;
  	inordertraversal(root->right);
  }
}

template <typename T>
void BinaryTree<T>::inorder()
{
  inordertraversal(root);
}

template <typename T>
typename BinaryTree<T>::Node* BinaryTree<T>::findNode(T input, Node* & root)
{
  if (root != NULL)
  { 
     if (root->value == input)
     {
    	return root;
     }
     else if (input > root->value)
     {
     findNode(input, root->right);
     }
     else
     {
     findNode(input, root->left);
     }
  }

  else
  return NULL;
}


template <typename T>
void BinaryTree<T>::deleteNode(T input)
{
  	Node* current = root;
  	Node* parent = NULL;

  	Node* foundNode = findNode(input, root);

  	if (foundNode == NULL)
  	{
  	cout << input << " is not in the Tree" << endl;
  	return;
  	}

  	else
  	{
  	 	while (current != NULL)
  		 {
		if (current->value < input)
		{
	  	parent = current;
	  	current = current->right;
		}
		else if (current->value > input)
		{
	  	parent = current;
		current = current->left;
		}
		else if (current->value == input)
		break;
    		}
  	}

  //current node being deleted has no left child. Reconnect parent to current node's right child
  if (current->left == NULL)
  {
    if (parent == NULL)
    parent = current->right;

    else
    {
	if (input > parent->value)
	parent->right = current->right;
	else
	parent->left = current->right;

	delete current;
    }


  }
   	//current node has a left child. Locate the rightmost node in the left subtree of current 		node
   	// and it's parent
  else
  {
    	Node* parentOfRightMost = current;
    	Node* rightMost = current->left;

    	//(Basically finding largest element in left subtree to replace current elements value
   	 while (rightMost->right != NULL)
   	{
		parentOfRightMost = rightMost;
		rightMost = rightMost->right;
    	}

   	//replace the element in current by the element in rightmost
   	current->value = rightMost->value;

	//if rightmost node has left subtree
   	if (parentOfRightMost->right == rightMost)
   	{
		parentOfRightMost->right = rightMost->left;
		delete rightMost;
   	}

	//special case
   	else
   	{
   		parentOfRightMost->left = rightMost->left;
   		delete rightMost;
   	}
  }
}

